panic when using wrong protocol against gateway port

- [x] Defect

To reproduce: 
 
```
$ echo '
"gateway": {
  "name": "do-lon1-kubecon-eu",
  "port": 4223,
}
' > conf/gateways.conf

# Run the server, then try to telnet and send:
# echo 'sub > 90' | nc 127.0.0.1 4223
# 
$ nats-server -c conf/gateways.conf 
[82250] 2019/05/22 09:15:29.094808 [INF] Starting nats-server version 2.0.0-RC12
[82250] 2019/05/22 09:15:29.094987 [INF] Git commit [not set]
[82250] 2019/05/22 09:15:29.095596 [INF] Gateway name is do-lon1-kubecon-eu
[82250] 2019/05/22 09:15:29.095781 [INF] Listening for gateways connections on 0.0.0.0:4223
[82250] 2019/05/22 09:15:29.096960 [INF] Address for gateway "do-lon1-kubecon-eu" is 172.16.48.18:4223
[82250] 2019/05/22 09:15:29.097596 [INF] Listening for client connections on 0.0.0.0:4222
[82250] 2019/05/22 09:15:29.097611 [INF] Server id is NADSGJ2DYU32GW3EFSIWA37KNYWFU4AFH6BBRGL3M4XHWY67FGYK2WUZ
[82250] 2019/05/22 09:15:29.097618 [INF] Server is ready
[82250] 2019/05/22 09:15:31.732195 [INF] 127.0.0.1:57115 - gid:1 - Processing inbound gateway connection
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1070ca6]

goroutine 22 [running]:
sync.(*Map).Load(0x0, 0x149aee0, 0xc000055d60, 0x1fc, 0x19264be, 0x1)
	/usr/local/Cellar/go/1.12/libexec/src/sync/map.go:103 +0x26
github.com/nats-io/nats-server/server.(*client).processGatewayRSub(0xc00022a000, 0xc000238204, 0x4, 0x1fc, 0x0, 0x0)
	/Users/wallyqs/repos/nats-dev/src/github.com/nats-io/nats-server/server/gateway.go:1716 +0x342
github.com/nats-io/nats-server/server.(*client).parse(0xc00022a000, 0xc000238200, 0xa, 0x200, 0xa, 0x0)
	/Users/wallyqs/repos/nats-dev/src/github.com/nats-io/nats-server/server/parser.go:405 +0x1acd
github.com/nats-io/nats-server/server.(*client).readLoop(0xc00022a000)
	/Users/wallyqs/repos/nats-dev/src/github.com/nats-io/nats-server/server/client.go:752 +0x1d5
github.com/nats-io/nats-server/server.(*Server).createGateway.func2()
	/Users/wallyqs/repos/nats-dev/src/github.com/nats-io/nats-server/server/gateway.go:737 +0x2a
created by github.com/nats-io/nats-server/server.(*Server).startGoRoutine
	/Users/wallyqs/repos/nats-dev/src/github.com/nats-io/nats-server/server/server.go:1911 +0x91
```

--- a/server/const.go
+++ b/server/const.go
@@ -40,7 +40,7 @@ var (
 
 const (
 	// VERSION is the current version for the server.
-	VERSION = "2.0.0-RC12"
+	VERSION = "2.0.0-RC14"
 
 	// PROTO is the currently supported protocol.
 	// 0 was the original

--- a/server/gateway.go
+++ b/server/gateway.go
@@ -149,6 +149,9 @@ type gateway struct {
 	infoJSON   []byte            // Needed when sending INFO after receiving INFO from remote
 	outsim     *sync.Map         // Per-account subject interest (or no-interest) (outbound conn)
 	insim      map[string]*insie // Per-account subject no-interest sent or modeInterestOnly mode (inbound conn)
+
+	// Set/check in readLoop without lock. This is to know that an inbound has sent the CONNECT protocol first
+	connected bool
 }
 
 // Outbound subject interest entry.
@@ -808,6 +811,15 @@ func (c *client) processGatewayConnect(arg []byte) error {
 		c.closeConnection(WrongGateway)
 		return ErrWrongGateway
 	}
+
+	// For a gateway connection, c.gw is guaranteed not to be nil here
+	// (created in createGateway() and never set to nil).
+	// For inbound connections, it is important to know in the parser
+	// if the CONNECT was received first, so we use this boolean (as
+	// opposed to client.flags that require locking) to indicate that
+	// CONNECT was processed. Again, this boolean is set/read in the
+	// readLoop without locking.
+	c.gw.connected = true
 
 	return nil
 }

--- a/server/parser.go
+++ b/server/parser.go
@@ -127,8 +127,17 @@ func (c *client) parse(buf []byte) error {
 
 		switch c.state {
 		case OP_START:
-			if b != 'C' && b != 'c' && authSet {
-				goto authErr
+			if b != 'C' && b != 'c' {
+				if authSet {
+					goto authErr
+				}
+				// If the connection is a gateway connection, make sure that
+				// if this is an inbound, it starts with a CONNECT.
+				if c.kind == GATEWAY && !c.gw.outbound && !c.gw.connected {
+					// Use auth violation since no CONNECT was sent.
+					// It could be a parseErr too.
+					goto authErr
+				}
 			}
 			switch b {
 			case 'P', 'p':

--- a/test/gateway_test.go
+++ b/test/gateway_test.go
@@ -446,3 +446,36 @@ func TestGatewaySendAllSubs(t *testing.T) {
 	clientSend("UNSUB 1\r\n")
 	gAExpect(runsubRe)
 }
+
+func TestGatewayNoPanicOnBadProtocol(t *testing.T) {
+	ob := testDefaultOptionsForGateway("B")
+	sb := runGatewayServer(ob)
+	defer sb.Shutdown()
+
+	for _, test := range []struct {
+		name  string
+		proto string
+	}{
+		{"sub", "SUB > 1\r\n"},
+		{"unsub", "UNSUB 1\r\n"},
+		{"rsub", "RS+ $foo foo 2\r\n"},
+		{"runsub", "RS- $foo foo 2\r\n"},
+		{"pub", "PUB foo 2\r\nok\r\n"},
+		{"anything", "xxxx\r\n"},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			// Create raw tcp connection to gateway port
+			client := createClientConn(t, ob.Gateway.Host, ob.Gateway.Port)
+			defer client.Close()
+			clientSend := sendCommand(t, client)
+			clientSend(test.proto)
+		})
+	}
+
+	// Server should not have crashed.
+	client := createClientConn(t, ob.Host, ob.Port)
+	defer client.Close()
+	clientSend, clientExpect := setupConn(t, client)
+	clientSend("PING\r\n")
+	clientExpect(pongRe)
+}

